File: 2019/08/solution.hs
#!/usr/bin/runghc

import Data.Char
import Data.List
import Data.Ord

main :: IO ()
main = do
    input <- readFile "input.txt"
    putStrLn $ task1 input
    putStrLn $ task2 input

task1 :: String -> String
task1 input =
    let
        (width, height) = (25, 6)
        layers = chunks (width * height) $ rstrip input
        layer = minimumBy (comparing $ countElem '0') $ layers
    in
        show $ (countElem '1' layer) * (countElem '2' layer)

task2 :: String -> String
task2 input =
    let (width, height) = (25, 6)
    in drawImage width $ decodeImage (width, height) $ rstrip input

rstrip :: String -> String
rstrip =
    reverse . dropWhile isSpace . reverse

countElem :: Eq a => a -> [a] -> Int
countElem x xs =
    length $ elemIndices x xs

chunks :: Int -> [a] -> [[a]]
chunks _ [] = []
chunks n xs =
    let (chunk, rest) = splitAt n xs
    in chunk : chunks n rest

decodeImage :: (Int, Int) -> String -> String
decodeImage (width, height) encodedImg =
    let
        layers = chunks (width * height) encodedImg
        pixel_layers = transpose layers
    in
        map (head . dropWhile (== '2')) pixel_layers

drawImage :: Int -> String -> String
drawImage width img =
    intercalate "\n" $ chunks width $ map color img

color :: Char -> Char
color '0' = 'X'
color '1' = '_'
color '2' = ' '

File: 2019/08/solution.py
#!/usr/bin/python3

width, height = 25, 6

def group_layers(data, width, height):
    image_size = width * height
    for i in range(0, len(data), image_size):
        yield data[i:i+image_size]

def count_occurrences(iterable, elem):
    return sum(i == elem for i in iterable)

def decode_image(data, width, height):
    for pixel in zip(*group_layers(data, width, height)):
        yield next(value for value in pixel if value != '2')

def draw_image(image, width):
    color = ('X', '_', ' ')
    for i, pixel in enumerate(image):
        yield color[int(pixel)]
        if (i+1) % width == 0:
            yield '\n'

def task1(input):
    image_layers = group_layers(input.strip(), width, height)
    layer = min(image_layers, key=lambda x: count_occurrences(x, '0'))
    return count_occurrences(layer, '1') * count_occurrences(layer, '2')

def task2(input):
    decoded = ''.join(decode_image(input.strip(), width, height))
    return ''.join(draw_image(decoded, width))

input = open('input.txt', 'r').read()
print(task1(input))
print(task2(input))
